pkg/opensearch/response_parser.go

```diff
@@ -86,29 +87,187 @@ func (rp *responseParser) getTimeSeries(configuredFields es.ConfiguredFields) (*
+			queryRes = processTraceSpansResponse(res, queryRes)
+		} else {
+			switch target.Metrics[0].Type {
+			case rawDataType:
+				queryRes = processRawDataResponse(res, configuredFields, queryRes)
+			case rawDocumentType:
+				queryRes = processRawDocumentResponse(res, target.RefID, queryRes)
+			case logsType:
+				queryRes = processLogsResponse(res, target.Metrics[0].Settings.Get("limit").MustString(), configuredFields, queryRes)
+			default:
+				props := make(map[string]string)
+				err := rp.processBuckets(res.Aggregations, target, &queryRes, props, 0)
+				if err != nil {
+					return nil, err
+				}
+				rp.nameFields(&queryRes.Frames, target)
+				rp.trimDatapoints(&queryRes.Frames, target)
 			}
-			rp.nameFields(&queryRes.Frames, target)
-			rp.trimDatapoints(&queryRes.Frames, target)
-		}
 
+		}
 		result.Responses[target.RefID] = queryRes
+
 	}
 	return result, nil
 }
 
-func processLogsResponse(res *es.SearchResponse, configuredFields es.ConfiguredFields, queryRes backend.DataResponse) backend.DataResponse {
+func processTraceSpansResponse(res *es.SearchResponse, queryRes backend.DataResponse) backend.DataResponse {
+	propNames := make(map[string]bool)
+	docs := make([]map[string]interface{}, len(res.Hits.Hits))
+
+	for hitIdx, hit := range res.Hits.Hits {
+		var withKeysToObj map[string]interface{}
+		if hit["_source"] != nil {
+			// some k:v pairs come from OpenSearch with field names in dot notation: 'span.attributes.http@status_code': 200,
+			// namely TraceSpanRow.Attributes and TraceSpanRow.Resource
+			// this turns everything into maps we can index and access
+			withKeysToObj = utils.FlattenNestedFieldsToObj(hit["_source"].(map[string]interface{}))
+		}
+
+		doc := map[string]interface{}{
+			"_id":     hit["_id"],
+			"_type":   hit["_type"],
+			"_index":  hit["_index"],
+			"_source": withKeysToObj,
+		}
+
+		for k, v := range withKeysToObj {
+			// some field names TraceView viz needs do not correspond to what we get from OpenSearch, this remaps them
+			if k == "startTime" {
+				startTime, err := utils.TimeFieldToMilliseconds(v)
+				if err != nil {
+					return backend.ErrDataResponse(500, err.Error())
+				}
+				doc[k] = startTime
+				continue
+			}
+			if k == "durationInNanos" {
+				value, isNumeric := v.(float64) // Check for float64
+				if isNumeric {
+					// grafana needs time in milliseconds
+					duration := value * 0.000001
+					doc["duration"] = duration
+					continue
+				} else {
+					backend.Logger.Debug("durationInNanos is not a float64")
+				}
+			}
+			if k == "parentSpanId" {
+				doc["parentSpanID"] = v
+				continue
+			}
+			if k == "traceId" {
+				doc["traceID"] = v
+				continue
+			}
+			if k == "spanId" {
+				doc["spanID"] = v
+				continue
+			}
+			if k == "name" {
+				doc["operationName"] = v
+				continue
+			}
+			if k == "resource" {
+				resourceAttributes, ok := v.(map[string]interface{})["attributes"].(map[string]interface{})
+				if resourceAttributes != nil && ok {
```

Comment:
It seems like we could separate out a utility function for converting the key values. A helper for the `events` would also help reduce the nesting and make it easier to read.